### PR TITLE
Harden and optimize stage finder

### PR DESCRIPTION
**Pull request by svanoort**
Avoids false positives on stage finding (found one that will be an issue for stage view, though normally that is desirable). 

Also avoids getting actions, which is MUCH slower than branching classcheck logic.

### CODE REVIEW

**Comment by svanoort** [comment]
Verified that this resolves a breakage of the one known stage view extension, testing against 1.651 and 2.7.3 builds.

**Comment by svanoort** [comment]
@reviewbybees

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java [hunk 1]
```diff
@@ -22,7 +26,19 @@ public boolean isStartInsideChunk() {
 
     @Override
     public boolean isChunkStart(@Nonnull FlowNode current, @CheckForNull FlowNode previous) {
-        return current.getAction(LabelAction.class) != null && current.getAction(ThreadNameAction.class) == null;
+        if ((current instanceof StepAtomNode || current instanceof StepStartNode) && !((((StepNode) current).getDescriptor()) instanceof StageStep.DescriptorImpl)) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: :bug: checking for `StageStep` incorrectly discriminates against other steps which add `LabelAction`, of which `stage` is just one example (that happens to do nothing else).

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java [hunk 2]
```diff
@@ -22,7 +26,19 @@ public boolean isStartInsideChunk() {
...
+            // Faster than looking at actions
+            return false;
+        } else if (current instanceof BlockEndNode) {
+            return false;
+        } else if (current instanceof StepStartNode) {
+            StepStartNode startNode = (StepStartNode)current;
+            if (!(startNode.getDescriptor() instanceof StageStep.DescriptorImpl)) {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: Ditto.

**Comment by jglick** [comment]
As for performance of `getAction`, see [JENKINS-30896](https://issues.jenkins-ci.org/browse/JENKINS-30896).

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java (hunk 1 above)
Comment (svanoort) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: @jglick There's a LabelChunkFinder in the workflow-api PR that covers the all-labels case. 

If you do not filter down to stages then checkpoints will not work correctly, unfortunately.

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: Not sure what that means exactly, but anyway the Javadoc does claim to be specific to be `stage`s.

File: src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java (hunk 2 above)
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/pipelinegraphanalysis/StageChunkFinder.java]: :ant: it seems that this case was already covered by the first clause.

**Comment by jglick** [comment]
:bee: though it looks a bit suspicious.

**Comment by svanoort** [comment]
WRT performance: benchmarking showed that getAction was the number 2 bottleneck. Ought to be cheap once the flownode is loaded (since both are loaded in the Tag) and IIRC actions should be cached after the first request to getActions( ) but still quite slow.

**Comment by svanoort** [comment]
For clarity: we need to distinguish stages explicitly from other labelled blocks/labels because stage view extensions may (and in the one key checkpoint case, *do*) depend on the getStageNodes(FlowExecution) and getStageNodes(FlowNode for stage start node) methods.  If you pick up checkpoints (which have a LabelAction but not ThreadNameAction) then they will not show within stages, since they will be their own chunks.  Thus stages will all show as having checkpoints in them.

**Pull request merged**